modules/core/arrow-typeclasses/src/main/kotlin/arrow/typeclasses/Decidable.kt

```diff
@@ -0,0 +1,368 @@
+          fc,
+          choose(
+            fd,
+            choose<E, Either<FF, G>, Either<E, Either<FF, G>>>(
+              fe,
+              choose(ff, fg, ::identity),
+              ::identity
+            ),
+            ::identity
+          ),
+          ::identity
+        ),
+        ::identity
+      ),
+      f
+    )
+
+  fun <A, B, C, D, E, FF, G, H, Z> choose(
+    fa: Kind<F, A>,
+    fb: Kind<F, B>,
+    fc: Kind<F, C>,
+    fd: Kind<F, D>,
+    fe: Kind<F, E>,
+    ff: Kind<F, FF>,
+    fg: Kind<F, G>,
+    fh: Kind<F, H>,
+    f: (Z) -> Either<A, Either<B, Either<C, Either<D, Either<E, Either<FF, Either<G, H>>>>>>>
+  ): Kind<F, Z> =
+    choose(
+      fa,
+      choose(
+        fb,
+        choose<C, Either<D, Either<E, Either<FF, Either<G, H>>>>, Either<C, Either<D, Either<E, Either<FF, Either<G, H>>>>>>(
+          fc,
+          choose(
+            fd,
+            choose<E, Either<FF, Either<G, H>>, Either<E, Either<FF, Either<G, H>>>>(
+              fe,
+              choose<FF, Either<G, H>, Either<FF, Either<G, H>>>(
+                ff,
+                choose(fg, fh, ::identity),
+                ::identity
+              ),
+              ::identity
+            ),
+            ::identity
+          ),
+          ::identity
+        ),
+        ::identity
+      ),
+      f
+    )
+
+  fun <A, B, C, D, E, FF, G, H, I, Z> choose(
+    fa: Kind<F, A>,
+    fb: Kind<F, B>,
+    fc: Kind<F, C>,
+    fd: Kind<F, D>,
+    fe: Kind<F, E>,
+    ff: Kind<F, FF>,
+    fg: Kind<F, G>,
+    fh: Kind<F, H>,
+    fi: Kind<F, I>,
+    f: (Z) -> Either<A, Either<B, Either<C, Either<D, Either<E, Either<FF, Either<G, Either<H, I>>>>>>>>
+  ): Kind<F, Z> =
+    choose(
+      fa,
+      choose(
+        fb,
+        choose<C, Either<D, Either<E, Either<FF, Either<G, Either<H, I>>>>>, Either<C, Either<D, Either<E, Either<FF, Either<G, Either<H, I>>>>>>>(
+          fc,
+          choose(
+            fd,
+            choose<E, Either<FF, Either<G, Either<H, I>>>, Either<E, Either<FF, Either<G, Either<H, I>>>>>(
+              fe,
+              choose(
+                ff,
+                choose<G, Either<H, I>, Either<G, Either<H, I>>>(
+                  fg,
+                  choose(fh, fi, ::identity),
+                  ::identity
+                ),
+                ::identity
+              ),
+              ::identity
+            ),
+            ::identity
+          ),
+          ::identity
+        ),
+        ::identity
+      ),
+      f
+    )
+
+  fun <A, B, C, D, E, FF, G, H, I, J, Z> choose(
+    fa: Kind<F, A>,
+    fb: Kind<F, B>,
+    fc: Kind<F, C>,
+    fd: Kind<F, D>,
+    fe: Kind<F, E>,
+    ff: Kind<F, FF>,
+    fg: Kind<F, G>,
+    fh: Kind<F, H>,
+    fi: Kind<F, I>,
+    fj: Kind<F, J>,
+    f: (Z) -> Either<A, Either<B, Either<C, Either<D, Either<E, Either<FF, Either<G, Either<H, Either<I, J>>>>>>>>>
```

Comment:
@raulraja is there any way of generalizing this for Product? Or convert between them?